sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/DigitalTwinMetadataJsonConverter.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Azure.DigitalTwins.Core
+{
+    /// <summary>
+    /// JSON converter to make it easier to deserialize a <see cref="BasicDigitalTwin"/>.
+    /// </summary>
+    internal class DigitalTwinMetadataJsonConverter : JsonConverter<DigitalTwinMetadata>
+    {
+        /// <inheritdoc/>
+        public override DigitalTwinMetadata Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType == JsonTokenType.Null)
+            {
+                return null;
+            }
+
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw new JsonException($"Unexpected token type {reader.TokenType} at index {reader.TokenStartIndex} Expected JsonTokenType.StartObject.");
```

Comment:
`JsonException` is the correct exception to throw here, in your custom converter.

However, if you prefer the deserializer to provide the diagnostic information for you, don't set the exception message.

Currently, this is the exception message you get:
```C#
JsonSerializer.Deserialize<DigitalTwinMetadata>("[]");
// Unhandled exception. System.Text.Json.JsonException: 
// Unexpected token type StartArray at index 0 Expected JsonTokenType.StartObject.
```

If you do:
```suggestion
                throw new JsonException();
```

You will get:
```C#
JsonSerializer.Deserialize<DigitalTwinMetadata>("[]");
// Unhandled exception. System.Text.Json.JsonException: 
// The JSON value could not be converted to Program+DigitalTwinMetadata. 
// Path: $ | LineNumber: 0 | BytePositionInLine: 1.
```

That is because the serializer catches `JsonException` thrown by anywhere in the callgraph, and then appends the "path" information before re-throwing it. It only appends this info if the exception message is null/empty:
https://github.com/dotnet/runtime/blob/3cc9f2c3ef4d7323de2c1fe74e86384de5b0c94e/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.ReadCore.cs#L92-L96
https://github.com/dotnet/runtime/blob/3cc9f2c3ef4d7323de2c1fe74e86384de5b0c94e/src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs#L343-L347

It's up to you to decide which one is more beneficial to the end user for diagnosing what went wrong.

Maybe there should be a way for the author of the converter to provided the "base" exception message, and then let the serializer append the extra info, always, but that isn't the current serializer behavior. The `AppendPathInformation` bool on `JsonException` is currently internal.

Reply:
Alternatively, **you could leave things as is**, so the user gets a more meaningful custom exception message string that you provided, and the caller can then choose to catch and access/output the `Path` and other properties on the `JsonException` to get more info, as needed, rather than relying on it being in the message string property directly.

```C#
// ...
catch (JsonException x)
{
    // Unexpected token type StartArray at index 0 Expected JsonTokenType.StartObject.
    // Path: $ | LineNumber: 0 | BytePositionInLine: 1.
    Console.WriteLine(ex.Message);
    Console.WriteLine($"Path: {ex.Path} | LineNumber: {ex.LineNumber} | BytePositionInLine: {ex.BytePositionInLine}.");
}
```

Reply:
I didn't know about the magic empty ctor that provided a good default. I think we'll go with that.